radix-engine/src/model/validated_transaction.rs

```diff
@@ -0,0 +1,399 @@
+use scrypto::core::SNodeRef;
+use scrypto::crypto::*;
+use scrypto::engine::types::*;
+use scrypto::prelude::ScryptoActor;
+use scrypto::rust::collections::{BTreeSet, HashMap};
+use scrypto::rust::string::String;
+use scrypto::rust::string::ToString;
+use scrypto::rust::vec;
+use scrypto::rust::vec::Vec;
+use scrypto::values::*;
+use crate::engine::{IdAllocator, IdSpace, Process};
+use crate::errors::RuntimeError;
+use crate::errors::RuntimeError::{ProofNotFound};
+use crate::ledger::SubstateStore;
+
+/// Represents a validated transaction
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ValidatedTransaction {
+    pub raw_hash: Hash,
+    pub instructions: Vec<ValidatedInstruction>,
+    pub signers: Vec<EcdsaPublicKey>,
+}
+
+impl ValidatedTransaction {
+    fn replace_ids(
+        mut values: Vec<ScryptoValue>,
+        proof_id_mapping: &mut HashMap<ProofId, ProofId>,
+        bucket_id_mapping: &mut HashMap<BucketId, BucketId>
+    ) -> Result<Vec<ScryptoValue>, RuntimeError> {
+        for value in values.iter_mut() {
+            value.replace_ids(proof_id_mapping, bucket_id_mapping)
+                .map_err(|e| match e {
+                    ScryptoValueReplaceError::BucketIdNotFound(bucket_id) => RuntimeError::BucketNotFound(bucket_id),
+                    ScryptoValueReplaceError::ProofIdNotFound(proof_id) => RuntimeError::ProofNotFound(proof_id),
+                })?;
+        }
+        Ok(values)
+    }
+
+    pub fn main<L: SubstateStore>(&self, proc: &mut Process<L>) -> (Vec<ScryptoValue>, Option<RuntimeError>) {
```

Comment:
Do you think it would be better to move this main to `TransactionValidator`? It doesn't belong to `ValidatedTransaction`.

Reply:
Either works